mongo/integration/unified/testrunner_operation.go

```diff
@@ -111,11 +137,94 @@ func executeTestRunnerOperation(ctx context.Context, operation *operation) error
 		coll := lookupString(args, "collectionName")
 		index := lookupString(args, "indexName")
 		return verifyIndexExists(ctx, db, coll, index, false)
+	case "loop":
+		var unmarshaledArgs loopArgs
+		if err := bson.Unmarshal(args, &unmarshaledArgs); err != nil {
+			return fmt.Errorf("error unmarshalling arguments to loopOptions: %v", err)
+		}
+		return executeLoop(ctx, &unmarshaledArgs, loopDone)
 	default:
 		return fmt.Errorf("unrecognized testRunner operation %q", operation.Name)
 	}
 }
 
+func executeLoop(ctx context.Context, args *loopArgs, loopDone <-chan struct{}) error {
+	// setup entities
+	entityMap := entities(ctx)
+	if args.errorsStored() {
+		if err := entityMap.addBSONArrayEntity(args.ErrorsEntityID); err != nil {
+			return err
+		}
+	}
+	if args.failuresStored() {
+		if err := entityMap.addBSONArrayEntity(args.FailuresEntityID); err != nil {
+			return err
+		}
+	}
+	if args.successesStored() {
+		if err := entityMap.addSuccessesEntity(args.SuccessesEntityID); err != nil {
+			return err
+		}
+	}
+	if args.iterationsStored() {
+		if err := entityMap.addIterationsEntity(args.IterationsEntityID); err != nil {
+			return err
+		}
+	}
+
+	for {
+		select {
+		case <-loopDone:
+			return nil
+		default:
+			var loopErr error
+			for i, operation := range args.Operations {
+				if operation.Name == "loop" {
+					return fmt.Errorf("loop sub-operations should not include loop")
+				}
+				loopErr = operation.execute(ctx, loopDone)
+
+				// if the operation errors, stop this loop
+				if loopErr != nil {
+					if !args.errorsStored() && !args.failuresStored() {
+						loopErr = fmt.Errorf("error running loop operation %v : %v", i, loopErr)
+						break
+					}
+					record := bson.D{
+						{"error", loopErr.Error()},
+						{"time", time.Now().Unix()},
+					}
+					errDoc, _ := bson.Marshal(record)
+					switch {
+					case !args.errorsStored():
+						entityMap.appendBSONArrayEntity(args.FailuresEntityID, errDoc)
+					case !args.failuresStored():
+						entityMap.appendBSONArrayEntity(args.ErrorsEntityID, errDoc)
+					// errors are test runner errors
+					// TODO GODRIVER-1950: use error types to determine error vs failure
```

Comment:
This comment should also explain why we currently use the "execution failed" substring to determine if `loopErr` is an error or a failure because it's not obvious where that's coming from.